programming-languages.md

```diff
@@ -1,51 +1,107 @@
+- move around the organisation
+- develop shared components
+- improve their personal development
+- master how they operate applications
 
-- when we're extending an existing codebase/ecosystem
-- when scripting in a particular environment
-- when we're experimenting during an alpha (with an expectation that it's replaced by something from the core set for beta)
+## Core languages
 
-The set of languages in the core group will change over time. This is likely
-to be through the adoption of a new language into the core group, and with
-an existing language being put on hold for new work.
+Our core languages are __Java, Python and Ruby__.
 
-## Languages in use not in the core group
+We've chosen these 3 languages because they are successfully used by
+teams at the moment, and we are confident in how to host and operate
+applications written in them. We'd advise you to carry out new
+development in one of them.
+
+## Languages in use in specific contexts
+
+### JavaScript
+
+This recommendation only applies to server-side JavaScript.
+The service manual has information on
+[using client-side JavaScript][manual_js].
+
+[manual_js]: https://www.gov.uk/service-manual/technology/using-progressive-enhancement
+
+The only way we're comfortable writing server-side JavaScript at the
+moment is with [Node.js][nodejs].
+
+[nodejs]: https://nodejs.org/
+
+Node.js is used by a few projects at GDS; GOV.UK Frontend, GOV.UK Pay and the Performance Platform.
+
+In the past we've had problems operating and maintaining products written in
+Node.js (for example maintaining dependencies, picking the appropriate
+libraries to use, understanding the appropriate application structure, and
+ensuring security of the code). We suspect that this is because:
+
+- It's younger and less mature than Java, Python and Ruby
+- We have fewer developers with Node.js skills than other tools we use
+
+Node.js is now frequently used in the wider software development ecosystem for
+building web application frontends. This means there are a lot of supporting
+tools and technologies and we expect this to gradually improve over time.
+
+It may therefore sometimes be appropriate to use this technology for rendering
+the web interface for our services. It is hard to set a specific boundary for
+where Node.js should be used, but our current approach is that Node.js shouldn't
+be used to implement any significant "application logic". GOV.UK Pay are adopting
+this approach by using Node.js only to create thin, client-facing applications
+which do not directly persist data (they may talk to an API for this); Node.js
+transforms an API to a web interface and back again. This divide matches reasonably
+well with the division between frontend and backend areas of specialism, and
+should allow the frontend to be developed with a reasonable level of decoupling
+from the backend applications.
+
+Teams that adopt Node.js must take steps to ensure that they have access to
```

Comment:
Rather than push this onto the teams, we should ask them to consult the forum before adopting node.js for any new work.